File: framec_tests/javascript/src/state_params/state_params.js
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

const FrameEvent = require("../framelang/FrameEvent")

class StateParams {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sInit_;
        this.#compartment = new StateParamsCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        this.param_log = [];
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    Next() {
        let e = FrameEvent("Next",null);
        this.#mux(e);
    }
    
    Prev() {
        let e = FrameEvent("Prev",null);
        this.#mux(e);
    }
    
    Log() {
        let e = FrameEvent("Log",null);
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sInit_:
                this.#sInit_(e);
                break;
            case this.#sSplit_:
                this.#sSplit_(e);
                break;
            case this.#sMerge_:
                this.#sMerge_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sInit_(e) {
        switch (e._message) {
            case "Next":
                {
                let compartment =  new StateParamsCompartment(this.#sSplit_);
                
                compartment.StateArgs["val"] = 1;
                
                this.#transition(compartment);
                
                return;
                }
                
        }
    }
    
    #sSplit_(e) {
        switch (e._message) {
            case "Next":
                {
                let compartment =  new StateParamsCompartment(this.#sMerge_);
                
                compartment.StateArgs["left"] = this.#compartment.StateArgs["val"];
                compartment.StateArgs["right"] = this.#compartment.StateArgs["val"] + 1;
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Prev":
                {
                let compartment =  new StateParamsCompartment(this.#sMerge_);
                
                compartment.StateArgs["left"] = this.#compartment.StateArgs["val"] + 1;
                compartment.StateArgs["right"] = this.#compartment.StateArgs["val"];
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Log":
                {
                this.got_param_do("val",(this.#compartment.StateArgs["val"]));
                
                return;
                }
                
        }
    }
    
    #sMerge_(e) {
        switch (e._message) {
            case "Next":
                {
                let compartment =  new StateParamsCompartment(this.#sSplit_);
                
                compartment.StateArgs["val"] = this.#compartment.StateArgs["left"] + this.#compartment.StateArgs["right"];
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Prev":
                {
                let compartment =  new StateParamsCompartment(this.#sSplit_);
                
                compartment.StateArgs["val"] = this.#compartment.StateArgs["left"] * this.#compartment.StateArgs["right"];
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Log":
                {
                this.got_param_do("left",(this.#compartment.StateArgs["left"]));
                this.got_param_do("right",(this.#compartment.StateArgs["right"]));
                
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    got_param_do(name,val) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    
    
};

//=============== Compartment ==============//

class StateParamsCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class StateParamsController extends StateParams {

	constructor() {
	  super()
	}
	got_param_do(name,val) {}
};

********************/

module.exports = StateParams

File: framec_tests/cpp/state_params/state_params.cpp
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

#include <vector>
#include <any>
using namespace std;
#include "../FrameLang/FrameLang.h"

//=============== Compartment ==============//

class StateParamsCompartment
{
public:
    int state;
    
    StateParamsCompartment(int state)
    {
        this->state = state;
    }
    
    std::unordered_map<std::string, std::any> stateArgs;
    std::unordered_map<std::string, std::any> stateVars;
    std::unordered_map<std::string, std::any> enterArgs;
    std::unordered_map<std::string, std::any> exitArgs;
    FrameEvent *_forwardEvent = nullptr;
};

class StateParams
{
private:
    StateParamsCompartment *_compartment_;
    StateParamsCompartment *_nextCompartment_;
    
    
    
public:
    StateParams()
    {
        // Create and intialize start state compartment.
        _state_ = static_cast<int>(StateParamsState::INIT);
        
        _compartment_ = new StateParamsCompartment(this->_state_);
        _nextCompartment_ = nullptr;
        
        
        // Send system start event
        FrameEvent frame_event(">", std::unordered_map<std::string, std::any>());
        _mux_(&frame_event);
    
    }
    
    // states enum
private:
    enum class StateParamsState
    {
        INIT = 0,
        SPLIT = 1,
        MERGE = 2
    };
    
    //====================== Multiplexer ====================//

private:
    void _mux_(FrameEvent *e)
    {
        if(this->_compartment_->state == static_cast<int>(StateParamsState::INIT))
        {
            this->_sInit_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(StateParamsState::SPLIT))
        {
            this->_sSplit_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(StateParamsState::MERGE))
        {
            this->_sMerge_(e);
        }
        
        
        if(this->_nextCompartment_ != nullptr)
        {
            StateParamsCompartment *nextCompartment = this->_nextCompartment_;
            this->_nextCompartment_ = nullptr;
            if (nextCompartment->_forwardEvent != nullptr && 
                nextCompartment->_forwardEvent->_message == ">")
            {
                this->_mux_(new FrameEvent( "<", this->_compartment_->exitArgs));
                this->_compartment_ = nextCompartment;
                this->_mux_(nextCompartment->_forwardEvent);
            }
            else
            {
                this->_doTransition_(nextCompartment);
                if (nextCompartment->_forwardEvent != nullptr)
                {
                    this->_mux_(nextCompartment->_forwardEvent);
                }
            }
            nextCompartment->_forwardEvent = nullptr;
        }
    }
    
        
        //===================== Interface Block ===================//
    public:
        void Next() {
            FrameEvent e("Next", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void Prev() {
            FrameEvent e("Prev", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void Log() {
            FrameEvent e("Log", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
    
    //===================== Machine Block ===================//

private:
    
    void _sInit_(FrameEvent *e)
    {
        if (e->_message == "Next") {
            StateParamsCompartment *compartment =  new StateParamsCompartment(static_cast<int>(StateParamsState::SPLIT));
            compartment->stateArgs["val"] = 1;
            
            this->_transition_(compartment);
            return;
        }
    }
    
    void _sSplit_(FrameEvent *e)
    {
        if (e->_message == "Next") {
            StateParamsCompartment *compartment =  new StateParamsCompartment(static_cast<int>(StateParamsState::MERGE));
            compartment->stateArgs["left"] = any_cast<int>(this->_compartment_->stateArgs["val"]);
            compartment->stateArgs["right"] = any_cast<int>(this->_compartment_->stateArgs["val"]) + 1;
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Prev") {
            StateParamsCompartment *compartment =  new StateParamsCompartment(static_cast<int>(StateParamsState::MERGE));
            compartment->stateArgs["left"] = any_cast<int>(this->_compartment_->stateArgs["val"]) + 1;
            compartment->stateArgs["right"] = any_cast<int>(this->_compartment_->stateArgs["val"]);
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Log") {
            got_param_do("val",any_cast<int>(this->_compartment_->stateArgs["val"]));
            return;
        }
    }
    
    void _sMerge_(FrameEvent *e)
    {
        if (e->_message == "Next") {
            StateParamsCompartment *compartment =  new StateParamsCompartment(static_cast<int>(StateParamsState::SPLIT));
            compartment->stateArgs["val"] = any_cast<int>(this->_compartment_->stateArgs["left"]) + any_cast<int>(this->_compartment_->stateArgs["right"]);
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Prev") {
            StateParamsCompartment *compartment =  new StateParamsCompartment(static_cast<int>(StateParamsState::SPLIT));
            compartment->stateArgs["val"] = any_cast<int>(this->_compartment_->stateArgs["left"]) * any_cast<int>(this->_compartment_->stateArgs["right"]);
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Log") {
            got_param_do("left",any_cast<int>(this->_compartment_->stateArgs["left"]));
            got_param_do("right",any_cast<int>(this->_compartment_->stateArgs["right"]));
            return;
        }
    }

//===================== Actions Block ===================//
    
    
    
public:
    void got_param_do(const std::string& name, int val)
    {
        param_log.push_back(name + "=" + std::to_string(val));
    }
    
    // Unimplemented Actions
    public:
    
    //===================== Domain Block ===================//
    
    
    public:
    std::vector<std::string> param_log ;
    
    
    //=============== Machinery and Mechanisms ==============//

private:
    int _state_;
    
    void _transition_(StateParamsCompartment *compartment)
    {
        _nextCompartment_ = compartment;
    }
    
    void _doTransition_(StateParamsCompartment *nextCompartment)
    {
        this->_mux_(new FrameEvent("<", _compartment_->exitArgs));
        this->_compartment_ = nextCompartment;
        this->_mux_(new FrameEvent(">", this->_compartment_->enterArgs));
    }
    
    public:
    string state_info(){
        return std::to_string(_compartment_->state);
        }
        
};

/********************

class StateParamsController : public StateParams
{
public:
	StateParamsController() : StateParams() {}
};

********************/

